projects/client/RabbitMQ.Client/src/client/impl/ConsumerWorkService.cs

```diff
@@ -1,84 +1,112 @@
-                if (key == null) { return; }
+                var newWorkPool = new WorkPool(model);
+                workPool = workPools.GetOrAdd(model, newWorkPool);
 
-                try
-                {
-                    foreach (var fn in actions)
-                    {
-                        fn();
-                    }
-                }
-                finally
+                // start if it's only the workpool that has been just created
+                if (newWorkPool == workPool)
                 {
-                    if (this.workPool.FinishWorkBlock(key))
-                    {
-                        var t = new Task(new Action(ExecuteThunk));
-                        t.Start(this.scheduler);
-                    }
+                    newWorkPool.Start();
                 }
             }
-            catch (Exception)
-            {
-#if NETFX_CORE
-                // To end a task, return
-                return;
-#else   
-                //Thread.CurrentThread.Interrupt(); //TODO: what to do?
-#endif
-            }
+
+            workPool.Enqueue(fn);
         }
 
-        public void AddWork(IModel model, Action fn)
+        public void StopWork(IModel model)
         {
-            if (this.workPool.AddWorkItem(model, fn))
+            WorkPool workPool;
+            if (workPools.TryRemove(model, out workPool))
             {
-                var t = new Task(new Action(ExecuteThunk));
-                t.Start(this.scheduler);
+                workPool.Stop();
             }
         }
 
-        public void RegisterKey(IModel model)
+        public void StopWork()
         {
-            this.workPool.RegisterKey(model);
+            foreach (var model in workPools.Keys)
+            {
+                StopWork(model);
+            }
         }
 
-        public void StopWork(IModel model)
+        class WorkPool
         {
-            this.workPool.UnregisterKey(model);
-        }
+            readonly ConcurrentQueue<Action> actions;
+            readonly AutoResetEvent messageArrived;
+            readonly TimeSpan waitTime;
+            readonly CancellationTokenSource tokenSource;
+            readonly string name;
 
-        public void StopWork()
-        {
-            this.workPool.UnregisterAllKeys();
+            public WorkPool(IModel model)
+            {
+                name = model.ToString();
+                actions = new ConcurrentQueue<Action>();
+                messageArrived = new AutoResetEvent(false);
+                waitTime = TimeSpan.FromMilliseconds(100);
+                tokenSource = new CancellationTokenSource();
+            }
+
+            public void Start()
+            {
+#if NETFX_CORE
+                Task.Factory.StartNew(Loop, TaskCreationOptions.LongRunning);
+#else
+                var thread = new Thread(Loop)
+                {
+                    Name = "WorkPool-" + name,
```

Comment:
I'd say that for profiling purposes it can quite helpful. For instancem if you use `Timeline` in dotTrace for instance you can exclude threads by their names narrowing the profiled area.